### PR TITLE
Implement (G)lweCiphertextList::from_fn/from_elem methods

### PR DESCRIPTION
**Pull request by leonardohn**
closes: #1108

### PR content/description

The first commit adds `from_fn` methods for `(G)lweCiphertextList`, enabling the user to directly initialize ciphertext lists using a closure. This closure must receive an `usize` as argument (the index) and return a scalar value that will initialize the list at the provided index. The closure is the last function parameter so that `rustfmt` can properly format the function call without skipping additional lines if the provided arguments can fit into a single line.

Following the name convention from `from_fn`, the second commit adds `from_elem` methods that behave similarly to the previous `new` methods, except for following a different parameter order to match `from_fn` parameters. Both `new` methods were marked as deprecated and internally invoke `from_elem` methods.

### Check-list:

* [ ] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] Relevant issues are marked as resolved/closed, related issues are linked in the description
* [ ] Check for breaking changes (including serialization changes) and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

**Comment by IceTDrinker** [comment]
hello @leonardohn as you can see in this bot comment https://github.com/zama-ai/tfhe-rs/pull/1129#issuecomment-2094777506

you first need to sign a CLA before we can take a look to your PR

Cheers

**Comment by IceTDrinker** [comment]
> @IceTDrinker I already signed the CLA, but the bot did not recognize for some reason.

Let me check on our end then

**Comment by IceTDrinker** [comment]
@cla-bot check

File: tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs [hunk 1]
```diff
@@ -150,7 +150,35 @@ pub type GlweCiphertextListView<'data, Scalar> = GlweCiphertextList<&'data [Scal
...
+    /// is a clone of the `fill_with` element.
+    ///
+    /// # Note
+    ///
+    /// This function is deprecated. Use [`GlweCiphertextList::from_elem`] instead.
+    #[inline(always)]
+    #[deprecated(
+        since = "0.6.1",
+        note = "use GlweCiphertextList::from_elem method instead"
+    )]
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs:163]: I would keep new, not deprecated it and not add the from_elem method

File: tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs [hunk 2]
```diff
@@ -161,19 +189,43 @@ impl<Scalar: UnsignedInteger> GlweCiphertextListOwned<Scalar> {
...
+    ///
+    /// # Note
+    ///
+    /// This function allocates a vector of the appropriate size and wraps it in the appropriate
+    /// type. If you want to encrypt data in the list you need to use
+    /// [`crate::core_crypto::algorithms::encrypt_glwe_ciphertext_list`] or a variant working on
+    /// a single ciphertext at a time [`crate::core_crypto::algorithms::encrypt_glwe_ciphertext`] on
+    /// the individual ciphertexts in the list.
+    ///
+    /// See [`GlweCiphertextList::from_container`] for usage.
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs:215]: I don't see a change in the from_container docstring to demonstrate usage ?

File: tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs [hunk 3]
```diff
@@ -154,17 +175,41 @@ impl<Scalar: UnsignedInteger> LweCiphertextListOwned<Scalar> {
...
+    ///
+    /// # Note
+    ///
+    /// This function allocates a vector of the appropriate size and wraps it in the appropriate
+    /// type. If you want to encrypt data you need to use
+    /// [`crate::core_crypto::algorithms::encrypt_lwe_ciphertext_list`] or its parallel variant
+    /// [`crate::core_crypto::algorithms::par_encrypt_lwe_ciphertext_list`] using this list as
+    /// output.
+    ///
+    /// See [`LweCiphertextList::from_container`] for usage.
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs:200]: I don't see an update to the from_container docstring to demonstrate usage ?

File: tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs [hunk 4]
```diff
@@ -143,7 +143,28 @@ pub type LweCiphertextListView<'data, Scalar> = LweCiphertextList<&'data [Scalar
...
+    ///
+    /// # Note
+    ///
+    /// This function is deprecated. Use [`LweCiphertextList::from_elem`] instead.
+    #[inline(always)]
+    #[deprecated(
+        since = "0.6.1",
+        note = "use LweCiphertextList::from_elem method instead"
+    )]
+    pub fn new(
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs:157]: I would keep new, not deprecated it and not add the from_elem method

**Review by IceTDrinker — CHANGES_REQUESTED** [review]
Just one small thing, I would keep the new method and not add from_elem, I don't see such a function in the rust std lib, the from_fn looks ok

Deprecation might be a bit overkill for now as we are not stable yet though it is certainly nice attention to detail.

**Comment by leonardohn** [comment]
> Just one small thing, I would keep the new method and not add from_elem, I don't see such a function in the rust std lib, the from_fn looks ok

I thought renaming the function would be a good addition as it describes how the container is being generated -- from a sample element, and it also justifies the parameter order swap without introducing breaking changes.

This functionality is available in the vec! macro, although it doesn't have a name there. Anyway, I added this in a separare commit so it would be easier to revert in that case.

File: tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs (hunk 2 above)
Comment (leonardohn) [reply tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs:208]: I added an usage example now.

File: tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs (hunk 3 above)
Comment (leonardohn) [reply tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs:192]: I added an usage example now.

**Comment by leonardohn** [comment]
@IceTDrinker that's right, it currently provides the raw index directly from the container, so the raw mask/body values can be manually set. I'm not sure if that's the best approach for the end user, though. If you agree, I can change `from_fn` to provide the list index only and then the user provides a full (G)lweCiphertext to initialize. What do you prefer?

**Comment by leonardohn** [comment]
In fact, to avoid double allocations I believe it is better to provide a mutable view of a (G)lweCiphertext, right?

**Comment by IceTDrinker** [comment]
For the list type I was thinking about it and likely having the current index inside the list and then the index of the given entity is probably better

**Comment by IceTDrinker** [comment]
> In fact, to avoid double allocations I believe it is better to provide a mutable view of a (G)lweCiphertext, right?

in what part of the code sorry ?

**Comment by leonardohn** [comment]
> in what part of the code sorry ?

There are four possibilities for the closure:

1. Provide the container index and receive a scalar value -- `Fn(usize) -> Scalar`;
    - This is the way it is implemented now and the simplest, not friendly for most of the users;
2. Provide the list and the entity index and receive a scalar value -- `Fn(usize, usize) -> Scalar`;
    - This is the way you proposed, it just requires applying div/mod over the container index;
3. Provide the list index and receive a owned ciphertext -- `Fn(usize) -> (G)lweCiphertextOwned<Scalar>`;
    - The most user friendly option but requires copying and double allocating;
4. Provide the list index and a mutable ciphertext reference -- `Fn(usize, (G)lweCiphertextMutView<Scalar>) -> ()`;
    - This avoids the problems above, but introduces mutability and the values must be default initialized;

Which one do you prefer?

**Comment by leonardohn** [comment]
> do you currently have a use case with lists ?

Yes. It's not doable in the current state of `tfhe-rs`, but I'm planning to upstream all the features I need. Just to contextualize, [this](https://arxiv.org/abs/2403.20190) is what I'm trying to implement.

> 3 and 4 feel like they should be already doable by first creating the list and using the iterators or getters to fill them as the users sees fit ?

That's right, I didn't consider this possibility. So I think we should stick with option 2.

**Comment by leonardohn** [comment]
> Thanks for sharing the use case, did you try Concrete-ML for this maybe ?

We introduced a few cryptographic primitives to achieve those results, so I believe it would be better to implement them over `tfhe-rs`, as they might benefit other general use cases. Despite having "neural networks" in the name, this algorithm does not involve traditional NN building blocks.

**Comment by IceTDrinker** [comment]
> > Thanks for sharing the use case, did you try Concrete-ML for this maybe ?
> 
> We introduced a few cryptographic primitives to achieve those results, so I believe it would be better to implement them over `tfhe-rs`, as they might benefit other general use cases. Despite having "neural networks" in the name, this algorithm does not involve traditional NN building blocks.

Understood !

Not sure about what we would put in TFHE-rs, is there a peer reviewed paper you could share on those ?

**Comment by leonardohn** [comment]
> Not sure about what we would put in TFHE-rs, is there a peer reviewed paper you could share on those ?

We are currently in process of peer-reviewing this, but I believe there are two major features that could benefit `tfhe-rs` in general:

- Secret `(G)lweCiphertextList` indexing based on a binary encoded `GgswCiphertextList`, which is a generalisation of the WoPBS procedure for ciphertext lists;
- Secret single-position `(G)lweCiphertextList` initialization based on a binary encoded `GgswCiphertextList`, using our proposed `CDEMUX` tree procedure -- this technique together with element-wise summation allows secret arbitrary modification of ciphertext lists;

Using these two features together, one is able to manipulate arbitrarily long ciphertext lists using ciphertext indices.

**Comment by IceTDrinker** [comment]
The doctests are failing run:

```
RUSTFLAGS="-C target-cpu=native" cargo +stable test --doc --profile devo         --features=x86_64-unix -p tfhe@0.6.0 -- from_container
```

File: tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs [hunk 5]
```diff
@@ -67,6 +67,26 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> GlweCiphertextList
...
+    ///     glwe_size,
+    ///     polynomial_size,
+    ///     ciphertext_count,
+    ///     ciphertext_modulus,
+    ///     |i, j| {
+    ///         // The `i` value represents the index in the list being filled;
+    ///         // The `j` value represents the index in the ciphertext being filled;
+    ///         // In this example, for every index pair `(i, j)`, we fill the
+    ///         // corresponding value using the formula `i + j`
+    ///         (i + j as u64)
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs:81]: likely fix is (i + j) as u64

File: tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs [hunk 6]
```diff
@@ -50,13 +50,31 @@ impl<Scalar: UnsignedInteger, C: Container<Element = Scalar>> LweCiphertextList<
...
+    /// let lwe_list = LweCiphertextList::from_fn(
+    ///     lwe_size,
+    ///     lwe_ciphertext_count,
+    ///     ciphertext_modulus,
+    ///     |i, j| {
+    ///         // The `i` value represents the index in the list being filled;
+    ///         // The `j` value represents the index in the ciphertext being filled;
+    ///         // In this example, for every index pair `(i, j)`, we fill the
+    ///         // corresponding value using the formula `i + j`
+    ///         (i + j as u64)
```
Comment (IceTDrinker) [inline tfhe/src/core_crypto/entities/lwe_ciphertext_list.rs:70]: same here

File: tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs (hunk 5 above)
Comment (leonardohn) [reply tfhe/src/core_crypto/entities/glwe_ciphertext_list.rs]: I forgot about the precedence, it should be working now.

**Review by IceTDrinker — APPROVED** [review]
Thanks !

**Pull request merged**